### ISSUE TITLE
Support pub packages

### ISSUE THREAD

**Issue by lukechurch** [dart-lang/dart-pad#901]
@devoncarew has started this

https://github.com/dart-lang/dartpad_server/pull/40

We need to consider:

-> Ensure that it can't be used as a DOS vector by triggering on the expensive package resolutions
-> Ensure that it plays nicely with intermittent failures of pub storage

**Comment by domesticmouse** [comment]
Blocked on https://github.com/dart-lang/dart-pad/issues/1128

**Comment by noahbetzen-wk** [comment]
In the meantime, is it possible to fork `dart-services`, hardcode a custom set of pub imports for local use, and self-host it for others to use? If so, where do the imports need to be done?

**Comment by domesticmouse** [comment]
Currently most of `dart-services` doesn’t know how to use packages, so there is no centralised place to whitelist packages.

**Comment by sergeytsivin** [comment]
Even if generic imports are not supported it would be helpful if 'package:test/test.dart' was implicitly imported by default, so that unit tests could be written directly in dartpad.

**Comment by rrousselGit** [comment]
Is there any plan to support this now that we have flutter support?

It'd be lovely to be able to write codelabs for 3rd party packages.
We could maintain a gist that concatenates all the files of a library - but that's not very easy to do for larger packages.

**Comment by startupgurukul** [comment]
Everything seems to have fallen in place for this to be resolved soon, looking forward to it.

**Comment by escamoteur** [comment]
Would be great, I just got stuck when I tried to use vector_math

**Comment by smkamranqadri** [comment]
Since dartpad launch and flutter interact, is there any plan for supporting external packages and multi file support?

**Comment by frank06** [comment]
The new Dartpad with Flutter support is great!

But I'd like to use [provider](https://pub.dev/packages/provider) in Dartpad, as it's an essential tool even included in the Flutter docs.

For starters, could "favorite" packages be whitelisted/allowed? https://flutter.dev/docs/development/packages-and-plugins/favorites

@RedBrogdon

**Comment by Dinver** [comment]
Need import package of pub.dev. Without it DartPad is useless.

**Comment by rrousselGit** [comment]
As a side-note, code-generators support would be great too. 

I wonder if we could have a `build_runner` running on the client

**Comment by domesticmouse** [comment]
@rrousselGit what do I need to do to enable code generators? I'm guessing we need to enable editing files other than `main.dart`.

**Comment by rrousselGit** [comment]
If we have the ability to write a `pubspec.yaml`, it should be feasible to detect if there's a dependency on `build_runner`, and then run `pub run build_runner build`

**Comment by mabinici** [comment]
Updates?

**Comment by domesticmouse** [comment]
This will probably be on my plate for next couple of quarters. First I have to land moving compilation from AppEngine Flex to Cloud Run, both for security and scalability concerns.

**Comment by angelhdz** [comment]
Anything yet? Without packages, I find myself all the time creating packages on the fly so I can help people with their Flutter projects, packages like provider, http, bloc/cubit, shared_preferences, etc. It isn't practical and it's time consuming. Thank you!

**Comment by insinfo** [comment]
Updates?

**Comment by marcusatiliusregulus** [comment]
i'd really hope they'd just whitelist a few packages like provider. maybe flutter favourites? i really like dartpad to be able to try some things while im on my phone commuting, but some essential packages are just missing...

ps: the http package to access https://jsonplaceholder.typicode.com would be nice too

**Comment by Chetan11-dev** [comment]
So could you latest know how could we add this feature to dart pad or any brainstorming map. You know it will be great for the next people was going to work on it.

**Comment by angelhdz** [comment]
Wow, nothing yet!?

**Comment by marcusatiliusregulus** [comment]
@Chetan11-dev i guess the same way you can import material.dart? or have some copies of the whitelisted packages on the server and provide a way to import them as if they were some local packages

**Comment by Chetan11-dev** [comment]
It will be interesting to know the source code of the server side implementation and see how could we add this feature. It seems to be a game changer feature you know.

**Comment by AzadCoder** [comment]
Hey flutter team, if you don't have time to add this main thing to the dartpad, You should hire more developers.
And I'm here. ;-)

**Comment by angelhdz** [comment]
Nothing yet!?

**Comment by srawlins** [comment]
Nothing yet. But stay tuned.

**Comment by AzadCoder** [comment]
I just want to let you know that I'm still waiting for this.

**Comment by chroline** [comment]
this issue has been open for 6 years, would love to see this feature implemented sometime soon 👍

**Comment by alisolanki** [comment]
I think this feature is very much needed in dartpad.
I was trying to implement package:intl and it just wouldn't work. Waiting to see a solution to this!

**Comment by nggepe** [comment]
I'm waiting for it 👀

**Comment by cver** [comment]
Me too  👀

**Comment by angelhdz** [comment]
I stopped using Dartpad. I can't wait for pub packages import feature to be added or at least the most used packages like _http, provider/riverpod, flutter_bloc, shared_preferences, path, uuid, rxdart, getX, 
image_picker, firebase,  etc._

So I now have a Flutter/Dart workspace in **VSCode** prepared to do quick projects on the fly, and that debugs in **Chrome**.

Had the dartpad devs made a statement about this kindly, at least, I would have waited more. But it's like they are ignoring us.

**Comment by Levi-Lesches** [comment]
@angelhdz, in the Dart team's defense, DartPad seems like more of a sandbox or demonstration environment, not one where you actually work on real projects. Having a workspace in a proper IDE is probably a _good_ thing here, even if it's for "quick" projects

**Comment by angelhdz** [comment]
> @angelhdz, in the Dart team's defense, DartPad seems like more of a sandbox or demonstration environment, not one where you actually work on real projects. Having a workspace in a proper IDE is probably a _good_ thing here, even if it's for "quick" projects

Silly me believing Dartpad was what codepen is for my JavaScript or what repl.it is for my Python. 🤷‍♂️

So, yeah, I will continue working offline.🤷‍♂️

**Comment by MihaMarkic** [comment]
> @angelhdz, in the Dart team's defense, DartPad seems like more of a sandbox or demonstration environment, not one where you actually work on real projects. Having a workspace in a proper IDE is probably a _good_ thing here, even if it's for "quick" projects

But still, if one wants to experiment, Dartpad could be the first choice. Not working on real projects, but working on issues and ideas. Not being able to import packages severely limits its use.

**Comment by RedBrogdon** [comment]
Hey, everybody. There's clearly a bunch of interest in this issue, so let me give y'all an update.

As you can see, we've had package support on our feature wish list for a while now, and over the last few months we've finally been able to allocate resources to implementing it. This isn't an easy process, since package support brings with it a number of concerns (security being one noteworthy example). As a result, it's something we have to be very careful about rolling out correctly.

That said, we're making good progress, and over the next couple months, we plan to deploy support for Dart and Flutter packages. This will start with a few Firebase services as well as packages created by the team at Google, then expand to include popular packages from the wider community. An allowlist will be used to gate which ones are available, and you can expect things to move on and off of that list over time as packages rise and fall in usage.

While I'm here, this seems like a great time to remind everybody of the three main tenets of Dart's [code of conduct](https://dart.dev/code-of-conduct):

* Respect people, their identities, their culture, and their work.
* Be kind. Be courteous. Be welcoming.
* Listen. Consider and acknowledge people's points before responding.

This is an open source project, which means its success depends on our ability to work together. Those three things are a big part of how we do that. 😄

**Comment by fredgrott** [comment]
Thanks, @RedBrogdon as some of us plan to use dartpad in our workflows to teach dart and flutter programming and it's a bit easier to teach cross-platform if I can import Flutter Programming Widgets library as it makes it somewhat easier

**Comment by lulupointu** [comment]
Thanks you for this update @RedBrogdon ! (Love you flutter youtube content btw :smile:)

I wanted to know if you could tell us more about the mechanism used for the allow-list. 
Will the package author have to apply or will it be based on pub points (or pub popularity)? Or something else?

I figure you might not know yet but just wanted to tell you that when you do it would be great to know (especially as a package author)!

Thanks again for the sneak peek into the flutter team work, I'm thrilled by the news and I hope to be able to use DartPad for interactive examples in the documentation of my package !

**Comment by startupgurukul** [comment]
Great news, also a expo go like client will make it a complete experience as browser is yet to catch up with what mobile apps can do

**Comment by insinfo** [comment]
@RedBrogdon 

This is great news, I would if possible to see the "mongo_dart" package as one of the first supported packages.

**Comment by cver** [comment]
Great, thank you so much, I believe this project will be one of must-known
flutter tool

>

**Comment by marcellodesales** [comment]
Any ETA?

**Comment by srawlins** [comment]
Nothing yet. But stay tuned.

**Comment by AzadCoder** [comment]
> Hey, everybody. There's clearly a bunch of interest in this issue, so let me give y'all an update.
> 
> As you can see, we've had package support on our feature wish list for a while now, and over the last few months we've finally been able to allocate resources to implementing it. This isn't an easy process, since package support brings with it a number of concerns (security being one noteworthy example). As a result, it's something we have to be very careful about rolling out correctly.
> 
> That said, we're making good progress, and over the next couple months, we plan to deploy support for Dart and Flutter packages. This will start with a few Firebase services as well as packages created by the team at Google, then expand to include popular packages from the wider community. An allowlist will be used to gate which ones are available, and you can expect things to move on and off of that list over time as packages rise and fall in usage.
> 
> While I'm here, this seems like a great time to remind everybody of the three main tenets of Dart's [code of conduct](https://dart.dev/code-of-conduct):
> 
> * Respect people, their identities, their culture, and their work.
> * Be kind. Be courteous. Be welcoming.
> * Listen. Consider and acknowledge people's points before responding.
> 
> This is an open source project, which means its success depends on our ability to work together. Those three things are a big part of how we do that. 😄

Thanks man, Yes we know this isn't an easy process, We are all programmers but you are the Google you should be able to do this. We will wait for that to make it as Google as possible as before.

**Comment by Levi-Lesches** [comment]
Well no official announcement yet but it looks like this has been implemented! Thanks!

![image](https://user-images.githubusercontent.com/20747538/125852087-6232ba5b-79b0-4c86-b315-131ab20cfade.png)

And clicking the little info button in the bottom corner gives this menu

![image](https://user-images.githubusercontent.com/20747538/125852145-434baf9b-76f0-4aa5-b675-c115cf745b7b.png)

Clicking on one of the packages takes you to its pub.dev page, and clicking on the version number goes to that version on pub.dev

**Comment by rockyoung** [comment]
I see the _http_ package has already listed on the menu which @Levi-Lesches was metioned above, but still it prompts error when I'm trying to import it:
 **Unsupported import(s): (package:http/http.dart)**
Did I miss something?

**Comment by srawlins** [comment]
This is not really launched yet, so the list of packages currently only reflects what has been resolved by the server's `pub get`. I think a small sampling of the visible list is directly importable, but I won't claim its sensical. :) Just soft-launched. I think the plan is to make everything in that list (and more) directly importable.

**Comment by hoc081098** [comment]
please add [`rxdart`](https://pub.dev/packages/rxdart)

**Comment by srawlins** [comment]
Launched. Quick turnaround on this one. 😀

**Comment by srawlins** [comment]
@hoc081098 please see https://github.com/dart-lang/dart-pad/wiki/Adding-support-for-a-new-package

**Comment by Faiizii** [comment]
please add "dart:isolate";

**Comment by domesticmouse** [comment]
@Faiizii Dart doesn't support isolates when cross compiling to JavaScript, which is how DartPad works.

**Comment by meysammahfouzi** [comment]
Please add [stacked](https://pub.dev/packages/stacked)

**Comment by domesticmouse** [comment]
> Please add [stacked](https://pub.dev/packages/stacked)

Please create a new issue. Commenting on a closed issue is just going to get lost in the mix.
